### ISSUE TITLE
Crash bug: range end index 55 out of range for slice of length 54

### ISSUE THREAD

**Issue by jamesnemesh** [noamteyssier/pipspeak#10]
Hi!

I appreciate that you've created this software - we're interested in pushing PipSeq data through our more standardized scRNASeq pipeline, and need to mush it into 10x format, so your code would be perfect.

However, I'm running into a crash bug.  I have a minimal set of 25 read pairs to reproduce:
```
 /home/unix/nemesh/bin/pipspeak --r1 r1.fastq.gz  --r2 r2.fastq.gz --config  /home/unix/nemesh/pipspeak/data/config_v3.yaml

Processed 0 reads                                                                                                                         
thread 'main' panicked at src/main.rs:92:34:
range end index 55 out of range for slice of length 54
stack backtrace:
⠒ Processed 0 reads                                                                                                                            0: rust_begin_unwind
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:595:5
   1: core::panicking::panic_fmt
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/panicking.rs:67:14
   2: core::slice::index::slice_end_index_len_fail_rt
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/slice/index.rs:76:5
   3: core::slice::index::slice_end_index_len_fail
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/slice/index.rs:68:9
   4: <core::ops::range::Range<usize> as core::slice::index::SliceIndex<[T]>>::index
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/slice/index.rs:411:13
   5: core::slice::index::<impl core::ops::index::Index<I> for [T]>::index
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/slice/index.rs:18:9
   6: pipspeak::parse_records::{{closure}}
             at /home/unix/nemesh/pipspeak/src/main.rs:92:34
   7: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/ops/function.rs:305:13
   8: core::option::Option<T>::map
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/option.rs:1075:29
   9: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/iter/adapters/map.rs:103:26
  10: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/iter/adapters/map.rs:103:9
  11: pipspeak::parse_records
             at /home/unix/nemesh/pipspeak/src/main.rs:111:40
  12: pipspeak::main
             at /home/unix/nemesh/pipspeak/src/main.rs:163:22
  13: core::ops::function::FnOnce::call_once
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Indexing past the end of the read feels like something we've all been through!

Please let me know if there's anything else I can do to help.  Unfortunately, I'm an R/Python/Java person, so no Rust debugging skills to offer.

You can replicate with the attached files:
[r2.fastq.gz](https://github.com/noamteyssier/pipspeak/files/12921709/r2.fastq.gz)
[r1.fastq.gz](https://github.com/noamteyssier/pipspeak/files/12921710/r1.fastq.gz)

**Comment by noamteyssier** [comment]
Hey @jamesnemesh 

I think the problem is that the reads in this case are sized to capture the expected length, but occasionally there is an offset of nucleotides at the beginning leading to a truncated UMI in the read. 

In the data I built this with I had plenty of buffer room on the right side of the reads that I never ran into issues of truncated UMIs so I didn't have an index check for that. I added that in though and also report the fraction of truncated UMIs that were filtered out in the log. 

Seemed to work for the test cases you attached, but let me know if you run into more issues.

Should be able to intsall the new update (0.1.9) with `cargo install pipspeak`
